File: src/main/java/se/kth/spork/spoon/conflict/IsUpperHandler.java
package se.kth.spork.spoon.conflict;

import se.kth.spork.util.Pair;
import spoon.reflect.declaration.CtElement;
import spoon.reflect.path.CtRole;
import spoon.reflect.reference.CtWildcardReference;

import java.util.Optional;

/**
 * A conflict handler for the IS_UPPER attribute. This appears on wildcards to specify if a type bound is an upper
 * or a lower type bound. For example <code><? extends String></code> means that IS_UPPER is true,
 * and <code><? super String></code> means that IS_UPPER is false.
 *
 * @author Simon Larsén
 */
public class IsUpperHandler implements ContentConflictHandler {

    @Override
    public CtRole getRole() {
        return CtRole.IS_UPPER;
    }

    @Override
    public Pair<Optional<Object>, Boolean> handleConflict(
            Optional<Object> baseVal,
            Object leftVal,
            Object rightVal,
            Optional<CtElement> baseElem,
            CtElement leftElem,
            CtElement rightElem) {
        return Pair.of(mergeIsUpper(baseElem, leftElem, rightElem), false);
    }

    private static Optional<Object> mergeIsUpper(Optional<CtElement> baseElem, CtElement leftElem, CtElement rightElem) {
        CtWildcardReference left = (CtWildcardReference) leftElem;
        CtWildcardReference right = (CtWildcardReference) rightElem;

        boolean leftBoundIsImplicit = left.getBoundingType().isImplicit();
        boolean rightBoundIsImplicit = right.getBoundingType().isImplicit();

        if (baseElem.isPresent()) {
            CtWildcardReference base = (CtWildcardReference) baseElem.get();
            boolean baseBoundIsImplicit = base.getBoundingType().isImplicit();

            if (leftBoundIsImplicit != rightBoundIsImplicit) {
                // one bound was removed, so we go with whatever is on the bound that is not equal to base
                return Optional.of(baseBoundIsImplicit == leftBoundIsImplicit ? left.isUpper() : right.isUpper());
            }
        } else {
            if (leftBoundIsImplicit != rightBoundIsImplicit) {
                // only one bound implicit, pick isUpper of the explicit one
                return Optional.of(leftBoundIsImplicit ? left.isUpper() : right.isUpper());
            }
        }

        return Optional.empty();
    }

}

File: src/main/kotlin/se/kth/spork/spoon/conflict/IsUpperHandler.kt
package se.kth.spork.spoon.conflict

import spoon.reflect.declaration.CtElement
import spoon.reflect.path.CtRole
import spoon.reflect.reference.CtWildcardReference

/**
 * A conflict handler for the IS_UPPER attribute. This appears on wildcards to specify if a type
 * bound is an upper or a lower type bound. For example `` means that
 * IS_UPPER is true, and `` means that IS_UPPER is false.
 *
 * @author Simon Larsén
 */
class IsUpperHandler : ContentConflictHandler {
    override val role: CtRole
        get() = CtRole.IS_UPPER

    override fun handleConflict(
        baseVal: Any?,
        leftVal: Any,
        rightVal: Any,
        baseElem: CtElement?,
        leftElem: CtElement,
        rightElem: CtElement
    ): Pair<Any?, Boolean> {
        return Pair(mergeIsUpper(baseElem, leftElem, rightElem), false)
    }

    companion object {
        private fun mergeIsUpper(
            baseElem: CtElement?,
            leftElem: CtElement,
            rightElem: CtElement
        ): Any? {
            val left = leftElem as CtWildcardReference
            val right = rightElem as CtWildcardReference
            val leftBoundIsImplicit = left.boundingType.isImplicit
            val rightBoundIsImplicit = right.boundingType.isImplicit
            if (baseElem != null) {
                val base = baseElem as CtWildcardReference
                val baseBoundIsImplicit = base.boundingType.isImplicit
                if (leftBoundIsImplicit != rightBoundIsImplicit) {
                    // one bound was removed, so we go with whatever is on the bound that is not equal
                    // to base
                    return if (baseBoundIsImplicit == leftBoundIsImplicit) left.isUpper else right.isUpper
                }
            } else if (leftBoundIsImplicit != rightBoundIsImplicit) {
                // only one bound implicit, pick isUpper of the explicit one
                return if (leftBoundIsImplicit) left.isUpper else right.isUpper
            }
            return null
        }
    }
}
